### ISSUE TITLE
Package validator should have a warning for the package which exposes types of its dependency with PrivateAssets="compile"

### ISSUE THREAD

**Issue by voroninp** [dotnet/sdk#38853]
If my package `Package` references package `Dependency` with `PrivateAssest="compile"`, it makes types from `Dependency` inaccessible for the client. 

However, if my package exposes types from `Dependency` as part of its public API, this contradicts the intent, so I'd expect [package validator](https://learn.microsoft.com/en-us/dotnet/fundamentals/apicompat/package-validation/overview) to emit a warning.
Unfortunately, it does not.

Related issue: github.com/NuGet/Home/issues/13246

**Comment by baronfel** [comment]
Thanks @KalleOlaviNiemitalo - I've applied those changes now and reopened.

**Comment by ericstj** [comment]
This does seem related to the type of work APICompat does already.  

We do examine references for types and would raise a warning about an unresolved reference.  Perhaps it's not surfacing correctly?  I can also imagine cases where a package is brand new and only has a single assembly - so there might not be any comparisons to do in which case we wouldn't resolve a dependency.

dotnet/runtime has similar testing in https://github.com/dotnet/runtime/blob/d88c7ba88627b4b68ad523ba27cb354809eb7e67/src/libraries/testPackages/build/packageTest.targets#L50
We consume the final package and test for closure of both reference and runtime assemblies.

**Comment by voroninp** [comment]
> I can also imagine cases where a package is brand new and only has a single assembly - so there might not be any comparisons to do in which case we wouldn't resolve a dependency.

Yes, no other versions. But there's no need to compare packages in this case, IMO.

**Comment by ViktorHofer** [comment]
@voroninp can you please share a minimal repro so that we can take a look? As @ericstj mentioned above, PackageValidation should already handle this scenarios as we only look at the references available (based on what msbuild / SDK tells us).

**Comment by voroninp** [comment]
@ViktorHofer [
[MyAwesomePackage.zip](https://github.com/dotnet/sdk/files/14983462/MyAwesomePackage.zip)
](url)

I get warnings neither for `dotnet publish` nor for `dot net pack`. I also do not get them when I call `apicompat package` tool:
> APICompat ran successfully without finding any breaking changes.
